debezium-server-redis/src/main/java/io/debezium/server/redis/RedisMemoryThreshold.java

```diff
@@ -26,55 +25,79 @@ public class RedisMemoryThreshold {
     private static final String INFO_MEMORY_SECTION_USEDMEMORY = "used_memory";
-
-    private static final Supplier<Boolean> MEMORY_OK = () -> true;
+    private static long accumulatedMemory = 0L;
+    private static long previouslyUsedMemory = 0L;
+    private static long totalProcessed = 0;
 
     private RedisClient client;
-
-    private int memoryThreshold;
-
-    private long memoryLimit;
-
-    private Supplier<Boolean> isMemoryOk;
+    private long memoryLimit = 0;
+    private long maximumMemory = 0;
 
     public RedisMemoryThreshold(RedisClient client, RedisStreamChangeConsumerConfig config) {
         this.client = client;
-        this.memoryThreshold = config.getMemoryThreshold();
         this.memoryLimit = 1024L * 1024 * config.getMemoryLimitMb();
-        if (memoryThreshold == 0 || memoryTuple(memoryLimit) == null) {
-            disable();
-        }
-        else {
-            this.isMemoryOk = () -> isMemoryOk();
-        }
     }
 
-    public boolean check() {
-        return isMemoryOk.get();
+    /**
+     * @param client
+     */
+    public void setRedisClient(RedisClient client) {
+        this.client = client;
     }
 
-    private boolean isMemoryOk() {
-        Tuple2<Long, Long> memoryTuple = memoryTuple(memoryLimit);
-        if (memoryTuple == null) {
-            disable();
+    /**
+     * Redis Enterprise samples the Redis database memory usage in an interval. Since the throughput of Redis is very big,
+     * it is impossible to rely on delayed memory usage reading to prevent OOM.
+     * In order to protect the Redis database and throttle down the sink we have created this mechanism:...
+     * @param extraMemory   - Estimated size of a single record.
+     * @param bufferSize    - Number of records in a batch.
+     * @param bufferFillRate - Rate in which memory can be filled.
+     * @return
+     */
+    public boolean checkMemory(long extraMemory, int bufferSize, int bufferFillRate) {
+        Tuple2<Long, Long> memoryTuple = memoryTuple();
+
+        maximumMemory = memoryTuple.getItem2();
+
+        if (maximumMemory == 0) {
+            totalProcessed += bufferSize;
+            LOGGER.debug("Total Processed Records: {}", totalProcessed);
             return true;
         }
-        long maxMemory = memoryTuple.getItem2();
-        if (maxMemory > 0) {
-            long usedMemory = memoryTuple.getItem1();
-            long percentage = usedMemory * 100 / maxMemory;
-            if (percentage >= memoryThreshold) {
-                LOGGER.info("Sink memory threshold percentage was reached. Will retry; " +
-                        "(current: {}%, configured: {}%, used_memory: {}, maxmemory: {}).",
-                        percentage, memoryThreshold,
-                        usedMemory, maxMemory);
-                return false;
-            }
+
+        long extimatedBatchSize = extraMemory * bufferFillRate;
+        long usedMemory = memoryTuple.getItem1();
```

Comment:
In the `memoryTuple()` function, the first element stored here could technically be `null` if the key for this element is not found and could result in a `NullPointerException`.  Similar to how you guarantee there is a non-null value for item2, you should do the same for item1.

Reply:
Fixed